jormungandr/src/secure/enclave.rs

```diff
@@ -38,14 +40,32 @@ impl Enclave {
         e
     }
 
+    pub async fn get_leader_id_if_present(&self, leader: &Leader) -> Option<LeaderId> {
+        match &leader.bft_leader {
+            Some(l) => {
+                let cache = self.added_cache.read().await;
+                cache.get(&l.sig_key.to_public().to_string()).cloned()
+            }
+            None => None,
+        }
+    }
     pub async fn get_leaderids(&self) -> Vec<LeaderId> {
         let leaders = self.leaders.read().await;
-        leaders.keys().map(|v| v.clone()).collect()
+        leaders.keys().cloned().collect()
     }
 
     pub async fn add_leader(&self, leader: Leader) -> LeaderId {
+        match self.get_leader_id_if_present(&leader).await {
+            Some(id) => return id,
+            None => {}
+        }
         let mut leaders = self.leaders.write().await;
         let next_leader_id = get_maximum_id(&leaders).next();
+        let mut cache = self.added_cache.write().await;
+        leader
+            .bft_leader
+            .as_ref()
+            .map(|l| cache.insert(l.sig_key.to_public().to_string(), next_leader_id));
```

Comment:
Same as [here](https://github.com/input-output-hk/jormungandr/pull/1760/files#r377229048), if it do not have a `bft_leader` from where we can take the public key then we cannot insert it in the already added cache.
I'm missing something here.